src/query/functions/linear/histogram_quantile.go

```diff
@@ -0,0 +1,376 @@
+		if err != nil {
+			// invalid bounds value for the bucket; drop it from the output.
+			continue
+		}
+
+		excludeTags := [][]byte{tags.Opts.MetricName(), tags.Opts.BucketName()}
+		tagsWithoutKeys := tags.TagsWithoutKeys(excludeTags)
+		id := tagsWithoutKeys.ID()
+		if buckets, found := bucketsForID[string(id)]; !found {
+			// Add a single indexed bucket for this ID with the current index only.
+			newBuckets := make([]indexedBucket, 0, initIndexBucketLength)
+			newBucket := indexedBucket{
+				upperBound: bound,
+				idx:        i,
+			}
+
+			newBuckets = append(newBuckets, newBucket)
+			bucketsForID[string(id)] = indexedBuckets{
+				buckets: newBuckets,
+				tags:    tagsWithoutKeys,
+			}
+		} else {
+			newBucket := indexedBucket{
+				upperBound: bound,
+				idx:        i,
+			}
+
+			buckets.buckets = append(buckets.buckets, newBucket)
+			bucketsForID[string(id)] = buckets
+		}
+	}
+
+	return bucketsForID
+}
+
+// sanitize sorts the bucket maps by upper bound, dropping any series which
+// have less than two buckets, or any that do not have an upper bound of +Inf
+func sanitizeBuckets(bucketMap bucketedSeries) {
+	for k, buckets := range bucketMap {
+		if len(buckets.buckets) < 2 {
+			delete(bucketMap, k)
+		}
+
+		sort.Sort(buckets)
+		maxBound := buckets.buckets[len(buckets.buckets)-1].upperBound
+		if !math.IsInf(maxBound, 1) {
+			delete(bucketMap, k)
+		}
+	}
+}
+
+func bucketQuantile(q float64, buckets []bucketValue) float64 {
+	// NB: some valid buckets may have been purged if the values at the current
+	// step for that series are not present.
+	if len(buckets) < 2 {
+		return math.NaN()
+	}
+
+	// NB: similar situation here if the max bound bucket does not have a value
+	// at this point, it is necessary to re-check.
+	if !math.IsInf(buckets[len(buckets)-1].upperBound, 1) {
+		return math.NaN()
+	}
+
+	rank := q * buckets[len(buckets)-1].value
+
+	bucketIndex := sort.Search(len(buckets)-1, func(i int) bool {
+		return buckets[i].value >= rank
+	})
+
+	if bucketIndex == len(buckets)-1 {
+		return buckets[len(buckets)-2].upperBound
+	}
+
+	if bucketIndex == 0 && buckets[0].upperBound <= 0 {
+		return buckets[0].upperBound
+	}
+
+	var (
+		bucketStart float64
+		bucketEnd   = buckets[bucketIndex].upperBound
+		count       = buckets[bucketIndex].value
+	)
+
+	if bucketIndex > 0 {
+		bucketStart = buckets[bucketIndex-1].upperBound
+		count -= buckets[bucketIndex-1].value
+		rank -= buckets[bucketIndex-1].value
+	}
+
+	return bucketStart + (bucketEnd-bucketStart)*rank/count
+}
+
+// Process the block
+func (n *histogramQuantileNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	bucketedSeries := gatherSeriesToBuckets(seriesMetas)
+
+	q := n.op.q
+	if q < 0 || q > 1 {
+		return processInvalidQuantile(q, bucketedSeries, meta, stepIter, n.controller)
```

Comment:
Seems kind of weird to split this out like this. Maybe just have one function called `processQuantile` and put the inf logic in there.

Reply:
If we do that, will need to do the comparison for each step